sdk/java/src/main/java/org/symbol/sdk/utils/ArrayHelpers.java

```diff
@@ -0,0 +1,303 @@
+package org.symbol.sdk.utils;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * Array helper for reading and writing fixed and variable size objects
+ */
+public class ArrayHelpers {
+
+	/**
+	 * Deeply compares two elements.
+	 *
+	 * @param lhs Left object to compare.
+	 * @param rhs Right object to compare.
+	 * @return 1 if lhs is greater than rhs; -1 if lhs is less than rhs; 0 if lhs and rhs are equal.
+	 */
+	public static <U extends Comparable<U>> int deepCompare(final U lhs, final U rhs) {
+		return lhs.compareTo(rhs);
+	}
+
+	/**
+	 * Deeply compares two array elements.
+	 *
+	 * @param lhs Left object list to compare.
+	 * @param rhs Right object list to compare.
+	 * @return 1 if lhs is greater than rhs; -1 if lhs is less than rhs; 0 if lhs and rhs are equal.
+	 */
+	public static <U extends Comparable<U>> int deepCompare(final U[] lhs, final U[] rhs) {
+		if (lhs.length != rhs.length)
+			return lhs.length > rhs.length ? 1 : -1;
+
+		for (int i = 0; i < lhs.length; ++i) {
+			final int compareResult = deepCompare(lhs[i], rhs[i]);
+			if (0 != compareResult)
+				return compareResult;
+		}
+
+		return 0;
+	}
+
+	private static <T extends Serializer, U extends Comparable<U>> List<T> readArrayImpl(final ByteBuffer buffer,
+			final Function<ByteBuffer, T> factory, final Function<T, U[]> accessor, final BiPredicate<Integer, ByteBuffer> shouldContinue) {
+		final List<T> elements = new ArrayList<>();
+		T previousElement = null;
+		int i = 0;
+		while (shouldContinue.test(i, buffer)) {
+			final T element = factory.apply(buffer);
+
+			if (0 >= element.getSize())
+				throw new IllegalStateException("element size has invalid size");
+
+			if (null != accessor && null != previousElement && 0 <= deepCompare(accessor.apply(previousElement), accessor.apply(element)))
+				throw new IllegalStateException("elements in array are not sorted");
+
+			elements.add(element);
+			previousElement = element;
```

Comment:
where is shift happening? in factory.apply?

Reply:
Yeah, this is one thing that's different from the Javascript SDK. The factory uses the ByteBuffer, which automatically shifts when reading a buffer. In some cases, the ByteBuffer will throw first instead of our checks when checking buffer limits.
I can change it if you want to have the swifts explicitly, but this is a little more efficient.

Reply:
it's ok, but maybe add a comment